File: assignments/src/ReorderDataInLogFiles.java
// Approach:
// 1. Split logs in 2 parts
// 2. check if they are digit log or letter log
// 3. If both are letter logs, further sort lexicographically, if again same, sort using identifiers.
// Time: O(mn*logn)
// Space: O(m*logn)

import java.util.Arrays;

class ReorderDataInLogFiles {
    public String[] reorderLogFiles(String[] logs) {
        Arrays.sort(logs, (str1, str2) -> {
            String[] strArr1 = str1.split(" ", 2);  // splits in 2: [let2, own kit dig]
            String[] strArr2 = str2.split(" ", 2);
            boolean isDigit1 = Character.isDigit(strArr1[1].charAt(0));
            boolean isDigit2 = Character.isDigit(strArr2[1].charAt(0));

            if (!isDigit1 && !isDigit2) {
                // both are letter logs
                int comp = strArr1[1].compareTo(strArr2[1]);    // +ve 0 -ve
                if (comp == 0) {
                    return strArr1[0].compareTo(strArr2[0]);
                }
                return comp;
            } else if (!isDigit1 && isDigit2) {
                // -ve number makes sure first one is before second one
                return -6;
            } else if (isDigit1 && !isDigit2) {
                // +ve number makes sure second one is before first one
                return 11;
            } else {
                return 0;   // no change
            }
        });
        return logs;
    }
}

File: ReorderDataInLogFiles.cpp
// Time Complexity - O(l*nlogn), l is the avg length of each log, n are the total number of logs.
// Space Copmlexity - O(l*k), l is the avg length of eaach log, k is the number of comparisions. Is this correct?

class Solution {
    pair<string, string> breakStr(const string& str) {
        pair<string, string> p;
        int spaceIdx = str.find(' ');
        p.first = str.substr(0, spaceIdx); 
        p.second = str.substr(spaceIdx + 1); 
        return p;
    }

public:
    vector<string> reorderLogFiles(vector<string>& logs) {
        stable_sort(logs.begin(), logs.end(), [&](const string& str1, const string& str2) {
            pair<string, string> s1 = breakStr(str1);
            pair<string, string> s2 = breakStr(str2);

            bool isDigit1 = isdigit(s1.second[0]);
            bool isDigit2 = isdigit(s2.second[0]);

            if (!isDigit1 && !isDigit2) {
                if (s1.second == s2.second) {
                    return s1.first < s2.first;
                } else {
                    return s1.second < s2.second;
                }
            } else if (!isDigit1 && isDigit2) {
                return true;  // str1 is a letter-log, str2 is a digit-log
            } else if (isDigit1 && !isDigit2) {
                return false;  // str1 is a digit-log, str2 is a letter-log
            } else {
                return false;  // str1 and str2 are both digit-logs
            }
        });

        return logs;
    }
};